Add schema for annotations

nlisgo opened the pull request.

nlisgo committed:
Add schema for annotations

giorgiosironi commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,49 @@
+        items:
+            type: object
+            properties:
+                uri:
```
does the `uri` work as an id? To uniquely characterize an annotation

nlisgo replied:
It will most certainly be unique. Here is an example:

```
https://hyp.is/36GLapojEeektWtTtv2WnA/elifesciences.org/inside-elife/77dd2808/taking-a-closer-look-screening-of-images
```
The hypothesis annotation id is contained within the url `36GLapojEeektWtTtv2WnA`.

thewilkybarkid replied:
Hmm, the slug shouldn't be part of the ID as renaming the item causes it to change. Can we declare our ID, which would be `blog-article/77dd2808` in this case?

nlisgo replied:
I agree. I have not introduced an id, I was unclear whether it was necessary since I am not exposing an endpoint to retrieve a single annotation.

thewilkybarkid replied:
Sounds like we don't want to use their URI at all, instead we can change this to an ID property.

giorgiosironi commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,49 @@
+                    title: Content
+                    type: array
+                    items:
+                        oneOf:
```
what's the rationale for including some types of blocks but not others? a comment or a link could help explain that

nlisgo replied:
I believe that these will be the only blocks needed to present all of the content for the annotations. This is due to the limited markup that is permitted within an annotation on hypothesis. As this is still experimental we could revise it once we have migrated all of the annotations and we know whether we should be supporting any other blocks.

thewilkybarkid commented on src/api.raml:
```diff
@@ -185,6 +185,37 @@ traits:
+    description: |
+        Annotations
+    uriParameters:
+        id:
```
That looks like the ID of an annotation. Instead, maybe we should have `by` as a query string parameter, and make it required for now (so if we ever allow lists by everyone, make it optional). (Or even by multiple authors, allow it to be multiple.)

nlisgo replied:
Will make this adjustment.

thewilkybarkid commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,49 @@
+        type: integer
+        minimum: 0
+    items:
+        title: Annotationss
```
Typo

thewilkybarkid commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,49 @@
+                    type: string
+                target:
+                    description: Target text for the annotation
+                    type: string
```
Will this ever contain HTML?

nlisgo replied:
I will check but I believe that hypothesis strip out the HTML.

thewilkybarkid commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,49 @@
+                    type: string
+                    format: uri
+                title:
+                    description: Title
```
Title of what? The annotation?

nlisgo replied:
This would be the title of the page that has been annotated.

thewilkybarkid commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,49 @@
+                          - $ref: ../blocks/mathml.v1.yaml
+                          - $ref: ../blocks/paragraph.v1.yaml
+                          - $ref: ../blocks/quote.v1.yaml
+                          - $ref: ../blocks/youtube.v1.yaml
```
`minItems: 1`?

nlisgo replied:
`minItems: 1` has been added.

nlisgo replied:
It has been removed again as content is optional.

thewilkybarkid commented on src/samples/annotation-list/v1/first-page.json:
```diff
@@ -0,0 +1,47 @@
+                },
+                {
+                    "type": "paragraph",
+                    "text": "https://cdn.elifesciences.org/style-guide-images/elife-full-color-horizontal.png"
```
Is `image` meant to be a possibility?

nlisgo replied:
This could be interesting. The images that are displayed in the annotations are only referenced, they must be hosted elsewhere. We only support displaying images served on a IIIF server. I will talk to @NickDuf about how we should present these on the profiles page.

giorgiosironi replied:
For what it's worth, it's insanely easy to add a new source to our IIIF that points to `cdn.elifesciences.org` or to somewhere similar.

thewilkybarkid replied:
So the images aren't hosted by Hypothesis? We might want to do what we do with Medium and download them, and connect Loris to Annotations. (Preserves the record of the image too, though I guess it would change/disappear through the Hypothesis client?)

nlisgo replied:
Yes, the hypothesis client would not use our images unless we edited the annotation and that would not be appropriate. The images in an annotation are stored as url's only. They could be hosted absolutely anywhere.

My recommendation for now is that we either present them as clickable links or that we convert them to `<img src="http://example/image.png" />`. Will chat to @NickDuf.

thewilkybarkid replied:
We already handle inline images, so that would work. But if they include something that's massive/disappears...

nlisgo replied:
@thewilkybarkid we cannot control what the size of the image is that is referenced. The hypothes.is client will attempt to render it regardless.

thewilkybarkid replied:
Depending on what the design request is, this should become HTML (eg `<img>` or `<a>`). Maybe one day we can integrate Hypothesis and IIIF...

nlisgo replied:
At the moment the designs are showing it displayed as a text link `<a>`.

nlisgo committed:
Set author filter as query parameter and set as experimental

nlisgo committed:
Set minItems for annotation content and improve description

nlisgo commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,50 @@
+              - uri
+              - title
+              - published
+              - content
```
I think I may need to make content optional because we may wish to displayed highlighted text without an annotation.

thewilkybarkid replied:
Ah, forgot that was a thing. Need to revert `content.minItems` as well.

nlisgo replied:
As content is optional, is it not ok to leave the `minItems` as `1`. It is my understanding that you could omit the property entirely but if it is present then it should contain at least `1` item.

thewilkybarkid replied:
Easier to make empty the same as not there.

thewilkybarkid commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,50 @@
+                    type: string
+                    format: uri
+                title:
+                    description: Title of the annotated document
```
As in the article title? We might want its type+ID too? Maybe we should include the snippet instead.

nlisgo replied:
It would be the title of whichever document that has been annotated: Article, blog article, labs post.

If we were to serve up the snippet of the content then this would require us to derive that from the target url which shouldn't be too much trouble but would require us to do a request for each annotation to the relevant service or to store the snippets for all content as part of the annotations service.

Having reviewed the designs for the annotations to be displayed on the profiles page, adding the snippets into the response would not add much value at this stage other than ensuring that the correct title of the annotated document was displayed. As things stand it would be the title of the document at the time that the annotation was recorded.

thewilkybarkid replied:
> As things stand it would be the title of the document at the time that the annotation was recorded.

Interesting, so different to the snippet then. We can ignore that for now, and add later if we need it.

nlisgo committed:
Add properties for id, targetUri, private and hasParent

nlisgo committed:
Adjust annotation-list sample

thewilkybarkid commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,62 @@
+                private:
+                    description: If the annotation should only be visible to the author
+                    type: boolean
+                    default: false
```
We should match what we did for profiles https://github.com/elifesciences/api-raml/blob/d2c08de4c7f3493cb8bbed9200b23f2cab0da226/src/model/profile.v1.yaml#L14-L15

thewilkybarkid commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,62 @@
+                    type: array
+                    items:
+                        oneOf:
+                          - $ref: ../blocks/list.v1.yaml
```
`v2` (not that we see the stuff that changed here)

thewilkybarkid commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,62 @@
+                    description: Target text for the annotation
+                    type: string
+                targetUri:
+                    description: URI of the annotated document
```
https://github.com/elifesciences/api-raml/pull/188#discussion_r146541874 is still relevant here.

nlisgo committed:
Address feedback

nlisgo committed:
typo

nlisgo commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,89 @@
+                          - $ref: ../blocks/quote.v1.yaml
+                          - $ref: ../blocks/youtube.v1.yaml
+                item:
+                    oneOf:
```
in order to introduce these snippets we are going to have to store them as part of the annotations service or perform an API request for each annotation, so we might need to rethink this? @giorgiosironi @thewilkybarkid

giorgiosironi replied:
How many annotations are we talking about in a single response?

Recommendations uses the approach of making several internal API requests, without cache, to compose a result. However, they are not that many, ~5 requests per API response tops.
Search is doing the same but with caching (I think?) or is it actually persisting in ElasticSearch the snippets? Will check.

Just to understand the whole context, where does the requirement of serving the snippets came from?

nlisgo replied:
Each annotation will have a snippet. So, if you request 20 annotations in a list then there will be up to 20 requests.

giorgiosironi replied:
Search is not caching any API call explicitly, the snippets are stored in ElasticSearch and overwritten on bus events. Which is what `annotations` would have to do, with a different persistence storage probably.

nlisgo replied:
We can get the title and url of the highlighted article and the Hypothesis API response which is all we need for the display but it doesn't seem right not to include the whole snippet.

nlisgo committed:
Change item snippet for document object

nlisgo commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,65 @@
+                    default: false
+                document:
+                    title: Annotated document
+                    type: object
```
This object is made up of values already available in a hypothesis response. It should give us everything we need to serve the designs.

giorgiosironi commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,65 @@
+                    type: string
+                access:
+                    $ref: ../misc/access-level.v1.yaml
+                hasParent:
```
is there a way to find out the `id` or something similar for the parent annotations? Rather than only knowing if it has one? This to promote reusability of this API with a low effort.

giorgiosironi replied:
All the other field names seem to match the other schemas in this repository

nlisgo replied:
In the annotation schema the `references` property lists the annotation id of all parents.

https://github.com/hypothesis/h/blob/75714a4829bf17cdb760b98829942b2646e60542/docs/_extra/api-reference/schemas/annotation-schema.json#L25-L30

I will introduce this array. The last entry in the array is the immediate parent and the first entry is the top level annotation or page note. If it is empty, it has no parents and is a top-level annotation or page note.

nlisgo committed:
Introduce parent annotation ID’s

nlisgo commented:
One thing that belongs here also is the profile id and display name of the author of the annotation. These values are returned in the Hypothesis search response but we want to avoid fleshing out the profile snippet because that will potentially require an extra API call for each annotation.

nlisgo committed:
set minItems of parents

giorgiosironi commented:
Makes sense given the naming `parents`, although an empty array is usually the "NULL value" of lists

nlisgo committed:
Require parents property but allow it to be empty

giorgiosironi reviewed (approved):
I have no further concerns

thewilkybarkid commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,66 @@
+                access:
+                    $ref: ../misc/access-level.v1.yaml
+                parents:
+                    description: Array of parent annotation ID's
```
`Array of` is redundant; no need for an apostrophe.

thewilkybarkid commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,66 @@
+              - parents
+              - document
+              - published
+              - updated
```
Not required?

thewilkybarkid commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,66 @@
+                    description: Highlighted text
+                    type: string
+                published:
+                    description: Publication date (UTC)
```
Of the annotation? Probably need a different term here.

thewilkybarkid commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,66 @@
+                    type: array
+                    items:
+                        type: string
+                document:
```
Not a fan of these, would much rather this API translated it into our concepts (in this case content item). But it does allow us to do what we need (and redirects should cover any changes), so I can live with it (for now).

thewilkybarkid commented on src/api.raml:
```diff
@@ -191,6 +191,40 @@ traits:
+        queryParameters:
+            by[]:
+                description: |
+                    Only return annotations by at least one of these authors.
```
As we're not going to be including author details, we should probably make this a single author, or it's a bit confusing?

nlisgo committed:
Introduce use-date for annotations

nlisgo committed:
Modifications to schema property descriptions

Alos modifies the published property to be created
Maye updated an optional property

nlisgo committed:
Amend a couple of descriptions

nlisgo committed:
We don’t need to make by required

nlisgo commented on src/api.raml:
```diff
@@ -191,6 +191,51 @@ traits:
+                description: |
+                    Only return annotations by this author.
+                type: string[]
+                required: false
```
I've set this to be optional now.

thewilkybarkid replied:
I didn't think we were going to list _all_ annotations? This has the same author problem (you don't know whose annotation it is).

nlisgo replied:
Right, I'll set it back to required.

nlisgo committed:
alter type to single value

nlisgo commented on src/model/annotation-list.v1.yaml:
```diff
@@ -0,0 +1,66 @@
+                    $ref: ../misc/access-level.v1.yaml
+                parents:
+                    description: Parent annotation IDs
+                    type: array
```
Could this just be `type: string[]`?

thewilkybarkid replied:
No (this is JSON Schema, not RAML).

nlisgo committed:
Adjust example to be single value

nlisgo committed:
by should be required for now
